### DISCUSSION TITLE
IPC between Shadow and external process

### DISCUSSION THREAD

**Discussion by nikitaborisov** [shadow/shadow#2155] [Q&A] [answered]
We're looking for a way to do IPC between a process simulated in Shadow and an external one so that we can run some computations on a GPU without having to figure out how to do that from within Shadow. Basically we want a Shadow process to collect some data, pass it to a non-Shadow process for computation, and then receive results back to be used within a simulation. 

Our current thinking is that we could have the process write some file in the shadow.data/hosts/ directory, which would then get picked up by an external process. This external process could then write a results file, which the simulated process could watch for, and then continue once the file is available. Does this sound like it would work? Are there any issues around this workflow? Is there a better approach we should explore?

Ideally, we'd also like to be able to pause the Shadow simulation while this external computation is going on. Perhaps sending a SIGSTOP to the Shadow process from the external one would work?

**Comment by sporksmith** [comment, marked as answer]
Have you tried running your GPU code under Shadow? I haven't done any GPU development, so I don't really know what this would look like at a syscall level. If it's ultimately operations on some GPU device file it *might* be feasible to work under Shadow. That's probably the best approach for long term support, but it's possible it'd require some work in shadow to support e.g. additional syscalls or fcntl operations.

Your idea to communicate through a file will probably work for now, but might not work in the future. I suppose inside the simulation you could "busy loop" waiting for data to be written to a file; and yes I think sending a SIGSTOP to the process that's spinning would stop it and effectively pause the simulation. 

I *think* it might be a little easier to use a named pipe (created from outside of Shadow, but opened as a regular file inside the simulation), the whole shadow process will get blocked when the process inside the simulation gets blocked waiting to read from the pipe. This approach also is likely to break in the future.

@stevenengler is more familiar with this part of the code; maybe he can confirm or has better ideas.

**Reply by stevenengler** [reply]
> If it's ultimately operations on some GPU device file it might be feasible to work under Shadow. That's probably the best approach for long term support, but it's possible it'd require some work in shadow to support e.g. additional syscalls or fcntl operations.

If it's CUDA, my understanding is that it just involves some mmaps and some ioctls on a device file. Shadow should pass any ioctl calls to the underlying OS file, so I don't see any issues there. If you can get an strace of the GPU program, that would probably help too.

> I think it might be a little easier to use a named pipe (created from outside of Shadow, but opened as a regular file inside the simulation), the whole shadow process will get blocked when the process inside the simulation gets blocked waiting to read from the pipe. This approach also is likely to break in the future.

You may run into weird behaviour with respect to blocking/non-blocking behaviour on fifos and regular files. We don't currently support blocking reads on files, so having Shadow wait for data might not work.

https://github.com/shadow/shadow/blob/cc0ec729d5ab40d9347373a79f36c82c4b14b8df/src/main/host/syscall/unistd.c#L107-L113

Maybe you could disable interposition while you wait for the file by calling the shim's preloaded `shim_swapAllowNativeSyscalls()` from within your program? Or maybe you could just disable interposition and make the GPU calls, then re-enable interposition? You'd have to be careful about what syscalls are run during that time. This isn't supported by Shadow either, but may be a workaround.

**Reply by sporksmith** [reply]
> If you can get an strace of the GPU program, that would probably help too.

+1

> We don't currently support blocking reads on files

Oops - yeah Shadow used to do blocking reads on file reads; it looks like that's been fixed.

> Maybe you could disable interposition while you wait for the file by calling the shim's preloaded shim_swapAllowNativeSyscalls() from within your program? Or maybe you could just disable interposition and make the GPU calls, then re-enable interposition? You'd have to be careful about what syscalls are run during that time. This isn't supported by Shadow either, but may be a workaround.

Ah, that's a good idea. I agree with your offline suggestion that maybe we should make this a supported "escape hatch".

@nikitaborisov Here's the function @stevenengler is referring to: https://github.com/shadow/shadow/blob/main/src/lib/shim/shim.h#L17

I think you'll need to link against shadow's shim library in your own program (shadow/build/src/lib/shim/libshadow-shim.so), or else create a dummy library that provides a no-op version of this function. (You can't just define the no-op version in your own program, since the call then wouldn't go through the PLT, so wouldn't be overridden by shadow's shim library).

**Reply by stevenengler** [reply]
@sporksmith I made an issue for this in #2157, hopefully containing most of what we discussed. Feel free to add any other thoughts to it as well.

**Reply by nikitaborisov** [reply]
Thanks! We did try using the GPU library (tensorflow) but the whole simulation would grind to a halt when we called it. I like the escape hatch solution, will report back once we try it out.
